### ISSUE TITLE
Upstream certificates not accepted

### ISSUE THREAD

**Issue by boonware** [oauth2-proxy/oauth2-proxy#1901]
## Expected Behavior
Authenticated requests should be forwarded to the upstream without a certificate error.

## Current Behavior

Users successfully authenticated with GitHub Enterprise receive a 500 error message in the browser. Server logs show the following message:
```
oauthproxy.go:823] Error redeeming code during OAuth2 callback: error performing request: Post "https://github.acme.com/login/oauth/access_token": x509: certificate signed by unknown authority
```

## Steps to Reproduce (for bugs)

1. Configuration file is as follows:
```
email_domains = "*"
whitelist_domains = ".acme.com"
provider = "github"
provider_display_name = "ACME GitHub Enterprise"
github_org = "myOrg"
github_team = "myTeam"
scope = "user:email"
login_url = "https://github.acme.com/login/oauth/authorize"
redeem_url = "https://github.acme.com/login/oauth/access_token"
validate_url = "https://github.acme.com/api/v3"
provider_ca_files = "/tmp/carootcert.pem,/tmp/caintermediatecert.pem"
```
2. The following environment variables are also set to appropriate values:
```
OAUTH2_PROXY_CLIENT_ID
OAUTH2_PROXY_CLIENT_SECRET
OAUTH2_PROXY_UPSTREAMS
OAUTH2_PROXY_REDIRECT_URL
OAUTH2_PROXY_COOKIE_SECRET
```
3. User makes a browser request to the OAuth Proxy path `https://oauth-proxy.acme.com/myUpstream`.
4. User is redirected to our GitHub and logs in successfully.
5. User is redirected back to OAuth Proxy and sees 500 error message in the browser, and the server logs the message above.

## Context

* The certificate from the GitHub Enterprise server is issued by DigiCert.
* The certificate from the upstream service is issues by my company's intermediate CA.
* As shown un the configuration file above, both the root and intermediate CA certs from my company are loaded using the variable `provider_ca_files`. Going by the name, it is not clear if this applies only to the provider, or also the upstreams? Is there a way to set CA files for the upstreams?

## Your Environment

- Version used: 7.4.0

**Comment by JoelSpeed** [comment]
We don't currently have a way to set the CA files for the upstream AFAIK, that said, you should be able to add them to the system roots and OAuth2 Proxy will pick them up from there. Otherwise I'd recommend looking into `pkg/upstream` and seeing if you can find a way to introduce the feature in there

**Comment by speedythesnailnpg** [comment]
This issue should not be closed. There is no reason that an application or service that takes certificates should not accept a ca cert (optionally).

**Comment by emsixteeen** [comment]
This looks like it might be addressed with #2237?

**Comment by tuunit** [comment]
> This looks like it might be addressed with #2237?

The issue is those certificates are meant to be used for the providers and yes extending the system trust store would allow for setting additional CAs inside the `provider-ca-file` flag intended for the upstreams but that seems like an unintended side effect of the implementation to me.

**Comment by speedythesnailnpg** [comment]
> > This looks like it might be addressed with #2237?
> 
> The issue is those certificates are meant to be used for the providers only and yes extending the system trust store might allow for setting additional CAs inside the `provider-ca-file` flag intended for the upstreams but that seems like an unintended side effect of the implementation to me. As the cert pool is only referenced inside the providers and the actual system trust store is not being extend this will most probably not fix this issue anyhow.
> 
> We either need to extend the config flags with a need list of CAs for the upstreams or change the current ca config flag to be used for providers and upstreams

I would agree with this. Were this a Java,  JavaScript, or Typescript app, I would be willing to help contribute, but I am completely unfamiliar with Go.

**Comment by emsixteeen** [comment]
> > This looks like it might be addressed with #2237?
> 
> The issue is those certificates are meant to be used for the providers only and yes extending the system trust store might allow for setting additional CAs inside the `provider-ca-file` flag intended for the upstreams but that seems like an unintended side effect of the implementation to me. As the cert pool is only referenced inside the providers and the actual system trust store is not being extend this will most probably not fix this issue anyhow.
> 
> We either need to extend the config flags with a need list of CAs for the upstreams or change the current ca config flag to be used for providers and upstreams

Looking at this closer it seems that this issue describes a TLS error for token redemption -- which should fall under the provider's TLS configuration/context, so #2237 might address this issue as well ... 

It doesn't look like the error described in the issue actually relates to upstreams, but I could be mistaken... If it is indeed related to upstreams, then #2237 might not actually help, and the config would need to be extended for upstreams specifically.

**Comment by tuunit** [comment]
Hi @emsixteeen,
that is exactly my point. I don't think #2237 would be of any help for upstreams. Either we will need a new implementation for a config flag `upstreams-ca` or changing the current implementation of `providers-ca` to `custom-ca` or `additional-ca` to be used for the handshake of providers and upstreams. I'll ask the other maintainers about their opinion.

**Comment by tuunit** [comment]
We decide to keep the upstream and provider CAs separate. Therefore, we will need to introduce a new config field `upstreamCA`

**Comment by remram44** [comment]
This is still a problem, please reopen

**Comment by remram44** [comment]
It's only inactive in the sense that maintainers haven't review the pull request.

**Comment by remram44** [comment]
Again, this is ONLY waiting for maintainers. Stop sending your bots at our issues and patches if you are inactive. This is just rude.

**Comment by remram44** [comment]
Again, this is ONLY waiting for maintainers. Stop sending your bots at our issues and patches if you are inactive. This is just rude.

**Comment by remram44** [comment]
Again, this is ONLY waiting for maintainers. Stop sending your bots at our issues and patches if you are inactive. This is just rude.

**Comment by JoelSpeed** [comment]
Left a review on the PR, I think it's nearly ready to go. Apologies for being absent most of the year, life events have been getting in the way of me being able to get to the project

**Comment by attilaolah** [comment]
> We decide to keep the upstream and provider CAs separate. Therefore, we will need to introduce a new config field `upstreamCA`

Are there any plans for this, or would it be OK to try and contribute this feature?

A "workaround" I've found is to use `ssl_upstream_insecure_skip_verify` but that's not very nice.

**Comment by remram44** [comment]
The workaround I use is an initContainer running busybox to add my CA to the bundle, it has been in production for a couple of years now and works well.

**Comment by remram44** [comment]
Review the pull request or call the bots off, thank you

**Comment by JoelSpeed** [comment]
> Review the pull request or call the bots off, thank you

I did review the pull request, my feedback has not been addressed or responded to.

The bot is doing its job correctly here IMO and reminding us all that we should double check in case something is waiting on us

We are waiting on @emsixteeen, but if they do not have time continue this, you are welcome to open your own PR and continue their work I would expect

**Comment by emsixteeen** [comment]
@JoelSpeed, I didn't realize something was waiting on me ... I'll have a look.

**Comment by JoelSpeed** [comment]
> @JoelSpeed, I didn't realize something was waiting on me ... I'll have a look.

No problem, very easy to miss a github notification here and there! Thanks
